cmd/src/search_stream.go

```diff
@@ -0,0 +1,413 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"net/url"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+	"github.com/sourcegraph/src-cli/internal/streaming"
+)
+
+var labelRegexp *regexp.Regexp
+
+func init() {
+	labelRegexp, _ = regexp.Compile("(?:\\[)(.*?)(?:])")
+}
+
+// streamHandler handles search requests which contain the flag "stream".
+// Requests are sent to search/stream instead of the GraphQL api.
+func streamHandler(args []string) error {
+	flagSet := flag.NewFlagSet("streaming search", flag.ExitOnError)
+	var (
+		display  = flagSet.Int("display", -1, "Limit the number of results that are displayed. Note that the statistics continue to report all results.")
+		apiFlags = api.StreamingFlags(flagSet)
+	)
+	if err := flagSet.Parse(args); err != nil {
+		return err
+	}
+
+	query := flagSet.Arg(0)
+
+	t, err := parseTemplate(streamingTemplate)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create request.
+	client := cfg.apiClient(apiFlags, flagSet.Output())
+	req, err := client.NewHTTPRequest(context.Background(), "GET", "search/stream?q="+url.QueryEscape(query), nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Accept", "text/event-stream")
+	if *display >= 0 {
+		q := req.URL.Query()
+		q.Add("display", strconv.Itoa(*display))
+		req.URL.RawQuery = q.Encode()
+	}
+
+	// Send request.
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error sending request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Process response.
+	err = streaming.Decoder{
```

Comment:
I think I'd personally pull some of this out to make it more readable. All of these callbacks require only the template (and stderr, see below), so you could use closures to make it flatter:

```go
func newDecoder(t *template.Template) streaming.Decoder {
	return streaming.Decoder{
		OnProgress: onProgress(t),
		// [...]
	}
}

func onProgress(t *template.Template) func(*streaming.Progress) {
	return func(progress *streaming.Progress) {
		// We only show the final progress.
		if !progress.Done {
			return
		}
		err = t.ExecuteTemplate(os.Stdout, "progress", progress)
		if err != nil {
			fmt.Printf("ExecuteTemplate: %s\n", err)
		}
		return
	}
}
```